laiaaa0/assignment4/Map.cpp

```diff
@@ -0,0 +1,106 @@
+#include "Map.h"
+
+int directions_x [] = {0,0,-1,1};
+int directions_y [] = {1,-1,0,0};
+
+Map::Map(const std::vector<std::vector<char> > & tiles) {
+  this->num_rows_ = tiles.size();
+  if (num_rows_ > 0){
+    this->num_columns_ = tiles[0].size();
+  }
+  else {
+    this->num_columns_ = 0;
+  }
+  this-> map_tiles_ = tiles;
+
+}
+
+Map::Map(const std::string & tiles_str) {
+  this->map_tiles_.clear();
+  std::vector<char> current_row;
+  for (int i = 0; i<tiles_str.size(); ++i){
+      if (tiles_str[i] == '-'){
+        this->map_tiles_.push_back(current_row);
+        current_row.clear();
+      }
+      else {
+        if (tiles_str[i]==ISLAND_T || tiles_str[i]==WATER_T)  {
+          current_row.push_back(tiles_str[i]);
+        }
+      }
+  }
+  this->num_rows_ = this->map_tiles_.size();
+  if (this->num_rows_ > 0){
+    this->num_columns_ = this->map_tiles_[0].size();
+  }
+  else {
+    this->num_columns_ = 0;
+  }
+}
+
+bool Map::InRange(int value, int min, int max) {
+    return value>=min && value<=max;
+}
+
+bool Map::IsInMap(int x, int y) {
+    return InRange(x, 0, this->num_rows_-1) && InRange(y, 0, this->num_columns_-1);
+}
+
+
+bool Map::ExpandNode(int x, int y, std::vector<std::vector<bool> > & visited) {
+    if (!IsInMap (x,y)) return false;
+    if (visited[x][y]) return false;
+    visited[x][y] = true;
+    if (this->map_tiles_[x][y] == WATER_T){
+        return false;
+    }
+    else if (this->map_tiles_[x][y] == ISLAND_T){
+        for (int i = 0; i < 4; ++i){
+                int next_x = x + directions_x[i];
+                int next_y = y + directions_y[i];
+                ExpandNode(next_x, next_y, visited);
+        }
+        return true;
+    }
+}
+
+std::pair<int,int> Map::GetNextUnvisited(int x, int y, const std::vector<std::vector<bool> > & visited) {
```

Comment:
It's a bit confusing (at least for me), using x for rows and y for columns, since Cartesian graphs are usually the other way around.

Also, do you need this function? I think it's easier to iterate over all cells on `FindNumberIslands`

Reply:
I think the code is more legible if separating the functions, it made more sense in my head! But maybe not...